From 7615ae1b9982bc7f7a55bc85cb9bd7f4beec3fe2 Mon Sep 17 00:00:00 2001
From: Clemens Backes <clemensb@chromium.org>
Date: Thu, 05 Dec 2024 19:41:22 +0100
Subject: [PATCH] Merged: [wasm] Fix comparison of canonical struct types

In addition to the field types we should also check the mutability.

R=jkummerow@chromium.org

Fixed: 382291459
(cherry picked from commit 3852cf8b5bceed6a76e4537fc0aa191f9ed672a3)

Change-Id: I3841b3cfda8e605d724e8695397837eba653ce76
Reviewed-on: https://chromium-review.googlesource.com/c/v8/v8/+/6086197
Auto-Submit: Jakob Kummerow <jkummerow@chromium.org>
Reviewed-by: Matthias Liedtke <mliedtke@chromium.org>
Commit-Queue: Jakob Kummerow <jkummerow@chromium.org>
Commit-Queue: Matthias Liedtke <mliedtke@chromium.org>
Cr-Commit-Position: refs/branch-heads/13.1@{#32}
Cr-Branched-From: 7998da66cb2883ef9734743857713b1194212d9a-refs/heads/13.1.201@{#1}
Cr-Branched-From: 5e9af2a913539cf67091def99b62f49afece6f56-refs/heads/main@{#96554}
---

--- a/v8/src/wasm/canonical-types.h
+++ b/v8/src/wasm/canonical-types.h
@@ -335,10 +335,15 @@
 
     bool EqualStructType(const CanonicalStructType& type1,
                          const CanonicalStructType& type2) const {
-      return std::equal(
-          type1.fields().begin(), type1.fields().end(), type2.fields().begin(),
-          type2.fields().end(),
-          std::bind_front(&CanonicalEquality::EqualValueType, this));
+      return
+          // Compare fields, including a check that the size is the same.
+          std::equal(
+              type1.fields().begin(), type1.fields().end(),
+              type2.fields().begin(), type2.fields().end(),
+              std::bind_front(&CanonicalEquality::EqualValueType, this)) &&
+          // Compare mutabilities, skipping the check for the size.
+          std::equal(type1.mutabilities().begin(), type1.mutabilities().end(),
+                     type2.mutabilities().begin());
     }
 
     bool EqualArrayType(const CanonicalArrayType& type1,
